libcrux-psq/src/ecdh_binder.rs

```diff
@@ -0,0 +1,310 @@
+
+    let aad = ts_ttl;
+    let initiator_dh_pk =
+        libcrux_ecdh::secret_to_public(libcrux_ecdh::Algorithm::X25519, initiator_dh_sk)
+            .map_err(|_| Error::CryptoError)?;
+    let aead_mac =
+        libcrux::aead::encrypt_detached(&initiator_key, &initiator_dh_pk, initiator_iv, aad)
+            .map_err(|_| Error::CryptoError)?;
+
+    Ok((
+        psk,
+        ECDHPsk {
+            encapsulation,
+            initiator_dh_pk,
+            aead_mac,
+            psk_ttl,
+            ts,
+        },
+    ))
+}
+
+/// Derive an ECDH-bound PSK from a PQ-PSK message.
+///
+/// Errors if the PQ-PSK message lifetime is elapsed.
+pub fn receive_ecdh_bound_psq(
+    receiver_pqsk: &PrivateKey,
+    receiver_pqpk: &PublicKey,
+    receiver_dh_sk: &X25519PrivateKey,
+    ecdh_psk_message: &ECDHPsk,
+    sctx: &[u8],
+) -> Result<Psk, Error> {
+    let ECDHPsk {
+        encapsulation,
+        initiator_dh_pk,
+        aead_mac,
+        psk_ttl,
+        ts,
+    } = ecdh_psk_message;
+    let ss_q = receiver_pqsk.derive_pq_psk(receiver_pqpk, encapsulation, sctx)?;
+    let initiator_dh_pk_bytes: [u8; 32] = initiator_dh_pk[0..32]
+        .try_into()
+        .map_err(|_| Error::InvalidPublicKey)?;
+    let initiator_dh_pk_point = libcrux_ecdh::X25519PublicKey(initiator_dh_pk_bytes);
+    let ss_dh = libcrux_ecdh::x25519_derive(&initiator_dh_pk_point, receiver_dh_sk)
+        .map_err(|_| Error::CryptoError)?;
+
+    // ikm = ss_q || ss_dh
+    let mut ikm = Vec::from(ss_q);
+    ikm.extend_from_slice(&ss_dh.0);
+
+    let prk = libcrux_hkdf::extract(libcrux_hkdf::Algorithm::Sha256, b"", ikm);
+
+    let psk: [u8; DH_PSK_LENGTH] = libcrux_hkdf::expand(
+        libcrux_hkdf::Algorithm::Sha256,
+        prk,
+        DH_PSK_CONTEXT,
+        DH_PSK_LENGTH,
+    )
+    .map_err(|_| Error::CryptoError)?
+    .try_into()
+    .expect("should receive the correct number of bytes from HKDF");
+
+    // NOTE: Make this a real cipherstate and pass it outside?
+    let (initiator_iv, initiator_key, _receiver_iv, _receiver_key) = derive_cipherstate(psk)?;
+
+    let ts_seconds = ts.as_secs();
+    let ts_subsec_millis = ts.subsec_millis();
+    let mut ts_ttl = ts_seconds.to_be_bytes().to_vec();
+    ts_ttl.extend_from_slice(&ts_subsec_millis.to_be_bytes());
+    ts_ttl.extend_from_slice(&psk_ttl.as_millis().to_be_bytes());
+
+    let aad = ts_ttl;
+    let initiator_dh_pk_decrypted =
+        decrypt_detached(&initiator_key, &aead_mac.1, initiator_iv, aad, &aead_mac.0)
+            .map_err(|_| Error::CryptoError)?;
+
+    // validate TTL
+    let now = SystemTime::now();
+    let ts_since_epoch =
+        Duration::from_secs(ts_seconds) + Duration::from_millis((ts_subsec_millis).into());
+    if initiator_dh_pk_decrypted != *initiator_dh_pk
+        || now
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .expect("now cannot be before UNIX_EPOCH")
+            - ts_since_epoch
```

Comment:
This may underflow. You should check that this can't happen in the condition before.